Changes for `stable-2.9.0` (#5174)
linkerd/linkerd2: CHANGES.md

# Changes

## stable-2.9.0

This release extends Linkerd's zero-config mutual TLS (mTLS) support to all TCP
connections, allowing Linkerd to transparently encrypt and authenticate all TCP
connections in the cluster the moment it's installed. It also adds ARM support,
introduces a new multi-core proxy runtime for higher throughput, adds support
for Kubernetes service topologies, and lots, lots more, as described below:

* Proxy
  * Performed internal improvements for lower latencies under high concurrency

>   * Reduced performance impact of logging , especially when the `debug` or

Reviewer: ```suggestion
  * Reduced performance impact of logging, especially when the `debug` or
```

    `trace` log levels are disabled

>   * Improved error handling for DNS errors encountered when discovering control
>     plane addresses, which can be common during installation, before all
>     components have been started, allowing linkerd to continue to operate
>     normally in HA during node outages

Reviewer: Nit: I feel like this reads a little odd with all the commas, maybe something like:

```suggestion
  * Improved error handling for DNS errors encountered when discovering control
    plane addresses; this can be common during installation before all
    components have been started, allowing linkerd to continue to operate
    normally in HA during node outages
```

Author: I'm also a big semicolon fan :slightly_smiling_face:

* Control Plane
  * Added support for [topology-aware service
    routing](https://kubernetes.io/docs/concepts/services-networking/service-topology/)
    to the Destination controller; when providing service discovery updates to
    proxies the Destination controller will now filter endpoints based on the
    service's topology preferences
  * Added support for the new Kubernetes
    [EndpointSlice](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/)
    resource to the Destination controller; Linkerd can be installed with
    `--enable-endpoint-slices` flag to use this resource rather than the
    Endpoints API in clusters where this new API is supported

* Dashboard
  * Added new Spanish translations (please help us translate into your
    language!)
  * Added new section for exposing multicluster gateway metrics

* CLI

>   * Renamed the `--addon-config` flag to `--config` to clarify this flag can be

Reviewer 2: I think this point is missing some words

    used

>   * Added fish shell completions to the `linkerd` command to set any Helm value

Reviewer 2: Oh, I found the missing words!

Author: :rofl: dang vim

* Multicluster

>   * Replaced the single `service-mirror` controller, with separate controllers

Reviewer: Nit:

```suggestion
  * Replaced the single `service-mirror` controller with separate controllers
```

    that will be installed per target cluster through `linkerd multicluster
    link`
  * Changed the mechanism for mirroring services: instead of relying on
    annotations on the target services, now the source cluster should specify
    which services from the target cluster should be exported by using a label
    selector
  * Added support for creating multiple service accounts when installing
    multicluster with Helm to allow more granular revocation
  * Added a multicluster `unlink` command for removing multicluster links

* Prometheus
  * Moved Linkerd's bundled Prometheus into an add-on (enabled by default); this
    makes the Linkerd Prometheus more configurable, gives it a separate upgrade

>     lifecycle from the rest of the control plane, and will allow users to

Reviewer: Change the tense here to match the other points

```suggestion
    lifecycle from the rest of the control plane, and allows users to
```

    disable the bundled Prometheus instance
  * The long-awaited Bring-Your-Own-Prometheus case has been finally addressed:
    added `global.prometheusUrl` to the Helm config to have linkerd use an
    external Prometheus instance instead of the one provided by default
  * Added an option to persist data to a volume instead of memory, so that

>     historical metrics are available when prometheus is restarted

Reviewer: ```suggestion
    historical metrics are available when Prometheus is restarted
```

  * The helm chart can now configure persistent storage and limits

* Other
  * Added a new `linkerd.io/inject: ingress` annotation and accompanying

>     `--ingress` flag to the `inject command, to configure the proxy to support

Reviewer: ```suggestion
    `--ingress` flag to the `inject` command, to configure the proxy to support
```

Author: Ah that's what was messing up my syntax coloring!
